File: Tree/preorder-inorder-postorder Traversal.cpp

// Preorder-Inorder-Postorder in one code
// T(n) = O(3N) and S(n) = O(3N)
class Solution{
public:
    vector<vector<int>>PreorderInorderPostorder(TreeNode *root){
        vector<vector<int>> output;
        if (!root) return output;
        stack<pair<TreeNode *, int>> st;
        vector<int> pre, in, post;
        st.push({root,1});
        while(!st.empty()){
            auto it = st.top();
            st.pop();
            
            // preorder Traversal
            if (it.second == 1){
                pre.push_back(it.first->val);
                it.second++;
                st.push(it);
                if(it.first->left){
                    st.push({it.first->left, 1});
                }
            }
            // Inorder Traversal
            if (it.second == 2){
                in.push_back(it.first->val);
                it.second++;
                st.push(it);
                if(it.first->right){
                    st.push({it.first->right, 1});
                }
            }

            // Postorder Traversal
            if (it.second == 3){
                post.push_back(it.first->val);
            }
        }
        output.push_back(pre);
        output.push_back(in);
        output.push_back(post);
        return output;
    }
};

File: Tree/preorder-inorder-postorder Traversal.py


# Preorder-Inorder-Postorder in one code
# T(n) = O(3N) and S(n) = O(3N)
class Solution:
    def PreorderInorderPostorder(self, root:[TreeNode]) -> List[int]:
        if root is None:
            return []
        preorder, inorder, postorder = [], [], []
        stack =[[root,1]]
        while stack:
            root = stack.pop()
            # for preorder traversal
            if root[-1] == 1:
                root[-1]+=1
                preorder.append(root[0].val)
                stack.append(root)
                if root[0].left:
                    stack.append([root[0].left,1])

            # for inorder traversal
            elif root[-1] == 2:
                root[-1]+=1
                inorder.append(root[0].val)
                stack.append(root)
                if root[0].right:
                    stack.append([root[0].right,1])

            # for postorder traversal
            elif root[-1] ==3:
                postorder.append(root[0].val)

        return(preorder,inorder,postorder)


# All DFS traversals (preorder, inorder, postorder) in 1 line
# T(n) = O(N) and S(n) = O(N)

def preorder(root):
  return [root.val] + preorder(root.left) + preorder(root.right) if root else []

def inorder(root):
  return  inorder(root.left) + [root.val] + inorder(root.right) if root else []

def postorder(root):
  return  postorder(root.left) + postorder(root.right) + [root.val] if root else []
    
      
        
